### ISSUE TITLE
Bulk uploading of resources

### ISSUE THREAD

**Issue by zdavis** [ManifoldScholar/manifold#434]

**Comment by tsmyre** [comment]
Here's a link to the spreadsheet we had drafted up to begin this conversation: https://docs.google.com/spreadsheets/d/1d9hKa3CMQh6lvPCEqLqQtiRlgKurNBMK0CpCaiRjjzA/edit?usp=sharing

**Comment by zdavis** [comment]
@tsmyre How important is the first row in this spreadsheet? I'd like to write an import mechanism that's not entirely dependent on google sheets. Ideally, I think whatever we write would work with .csv data as well. The sheet you linked to above includes merged cells in that top row, which I don't think are really supported in csv. It also includes an instructional row that the import mechanism will have to know about. I worry that the logic here—skip row one, grab column headers from row two, then skip row three and start importing in row four—is too tailored to this particular spreadsheet and will make it difficult for other presses to utilize this feature unless they have that exact structure to start from. 

I'd like to see us use a simpler structure. The first row could be headers, and the subsequent rows could contain the actual data. If we do this, we could easily allow users to upload from different formats (xml, json, etc.), rather than having the logic bound tightly to this particular spreadsheet implementation.

But I'm also sensitive to what you're trying to accomplish here, which is to make the data entry process easier for authors by providing them with instructions. Could we break that instructional content out into our documentation?

Let me know what you think.

--Zach

**Comment by tsmyre** [comment]
@zdavis The first row was more for cosmetic organization than anything else, so I removed it. As you noted though it would be nice to have the instructions on the sheet (as well as in the documentation) for the sake of immediacy and easy. But they don't have to go below the column heads. I'm not sure if that helps with the logic, but I deleted what was the first row, replacing it with the instructions, so Row 1, Instructions; Row 2, Headers; Row 3, Data. I also removed all merged cells. Does that get at it, or does this require a bit more thinking through?

**Comment by zdavis** [comment]
It's time to get started on this for the v0.4.0 release. First step is to get design for this feature from Lael. My attempt at a design brief for this feature is as follows:

**Summary**
Authors and/or Press users need to be able to bulk upload resources to a project. This will take place in the Manifold backend, in the resources section of a project. They can do this by either providing the URL of a google spreadsheet or by uploading a CSV file. We'll call this the "source" for the bulk import. Once the source has been specified, Manifold will analyze the source and extract the column headers. Either Manifold or the user will need to map the columns in the source file to resource attributes. Once this mapping is complete, the user can proceed and the records will be imported. When the import has finished, Manifold needs to provide feedback about the results of the import.

**UI Requirements**

The UI is roughly broken up to accomodate the three things the user needs to be able to do: selecting the source, mapping the columns, and viewing the import results.

1. Source selection
	1. The user is given a choice between entering the URL of a google doc and uploading a CSV file. 
	1. The google doc choice is only visible if the Manifold installation has google API keys in place. If they're not, the user isn't given a choice and can _only_ upload a CSV file.
	1. There is a checkbox labelled "First row contains titles" — the import needs to be told if the columns have headers or not.
	1. After entering the source, the user needs to click "upload" or "proceed" to make Manifold fetch and process the source document.
1. Column mapping
	1. Manifold needs to show the user a list of all the columns in the document. If the "first row contains titles" checkbox was checked, then each item in the list will be the column header. If that was not checked, we show the column number and perhaps the value in the first row.
	1. Manifold needs to also show a list of all the available resource attributes. Some or all of the available attributes need to be mapped to the columns in the source document. This is typically done through a drag/drop interface. Some of the attributes are requird, for example "title" and "kind". The user can't proceed until these have been mapped to columns.
	1. There needs to be an "auto mapping" button. If this is pressed, Manifold will attempt to automatically map attributes to columns based on the column names.
	1. Once the mapping is complete, the user presses an "import" button to finalize the import.
1. Results
	1. After the import is done, the user needs to be given some information about what happened.
	1. We want to tell the user how many resources were updated and how many resources were created.
	1. It's possible that some resources can't be imported. We need to include error messages for each of these. For example: "row 15 could not be imported due to errors: missing title, missing kind"

**Recommendations**

Importing from CSV and mapping is pretty much a solved problem. We want to keep the UI as simple as possible, while recognizing that this is a somewhat complex, expert task, and will require some knowledge. I would avoid a staged approach, in which the user has to move between the steps. It would be better to show all three of the components in one view, although that view could live in a modal backend interface (similar to ingestion).

I'm attaching some screenshots from the October CMS that are relevant. October provides CSV import interfaces that are pretty easy to work with and could serve as a starting point or example for your design. We do not need the "file format" option that you see in the October UI.

The following image shows the October CSV import before the user has uploaded a source file.

<img width="909" alt="screen shot 2017-10-12 at 9 55 50 am" src="https://user-images.githubusercontent.com/255909/31508878-52a86500-af34-11e7-835a-8aea3a408259.png">

This next image shows what the UI looks like after the CSV file has been uploaded. As you can see, the list on the left is now populated with the columns from the source document. The user can then drag the available attributes from the list on the right and line them up with the columns. This drag/drop interface is appropriate, because it prevents the user from using an attribute twice. 

<img width="906" alt="screen shot 2017-10-12 at 9 55 31 am" src="https://user-images.githubusercontent.com/255909/31508890-5a0c5522-af34-11e7-8b9e-bb95973b9637.png">

**Comment by tsmyre** [comment]
@zdavis Do you need any input from me at the moment regarding Database fields or other, or is that better to tackle that in person in NYC next week?

**Comment by tsmyre** [comment]
As part of the general UX component of this issue, possible to show more than five projects at a time. Also, it might be handy to have the Featured slider show up on the Showing Projects page instead of within a project.
